// mv-file/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  FileMoverError: () => FileMoverError,
  createFileMover: () => createFileMover,
  moveFile: () => moveFile
});
module.exports = __toCommonJS(src_exports);
var import_fs_extra = __toESM(require("fs-extra"));
var import_pathe = __toESM(require("pathe"));
var import_fast_glob = __toESM(require("fast-glob"));
var import_events = require("events");
var FileMoverError = class extends Error {
  constructor(message, code, source, target, originalError) {
    super(message);
    this.code = code;
    this.source = source;
    this.target = target;
    this.originalError = originalError;
    this.name = "FileMoverError";
  }
};
var FileMover = class extends import_events.EventEmitter {
  constructor(options = {}) {
    super();
    this.options = {
      cwd: process.cwd(),
      base: "",
      dest: "",
      force: false,
      clean: false,
      verbose: false,
      concurrency: 4,
      ...options
    };
  }
  _emit(event, ...args) {
    return super.emit(event, ...args);
  }
  /**
   * 复制文件或目录
   * @param source 源文件路径
   * @param target 目标路径
   * @param isDir 源路径是否为目录
   * @param subPath 附加路径(相对base的路径)
   */
  async copyFile(source, target, isDir, subPath) {
    try {
      let finalTarget = target;
      const { force } = this.options;
      if (isDir) {
        if (subPath) finalTarget = import_pathe.default.join(target, subPath);
      } else {
        const sourceExt = import_pathe.default.extname(source);
        const targetExt = import_pathe.default.extname(target);
        if (!targetExt && sourceExt !== targetExt) {
          finalTarget = subPath ? import_pathe.default.join(target, subPath) : import_pathe.default.join(target, import_pathe.default.basename(source));
        }
      }
      this._emit("copy:start", source, finalTarget);
      await import_fs_extra.default.copy(source, finalTarget, {
        overwrite: force,
        errorOnExist: !force,
        preserveTimestamps: true
      });
      this._emit("copy:done", source, finalTarget);
    } catch (err) {
      throw new FileMoverError(`Failed to copy file`, "COPY_ERROR", source, target, err);
    }
  }
  /**
   * 清理空目录
   * @param {string} dir - 清理目录
   * @param {string} baseDir - 基准目录
   */
  async cleanEmptyDir(dir, baseDir = "") {
    try {
      if (!dir.startsWith(baseDir)) return;
      const items = await import_fs_extra.default.readdir(dir);
      if (items.length > 0) return;
      await import_fs_extra.default.rmdir(dir);
      const parentDir = import_pathe.default.dirname(dir);
      if (parentDir !== baseDir) {
        await this.cleanEmptyDir(parentDir, baseDir);
      }
    } catch (err) {
      if (err.code === "ENOENT") {
        if (this.options.verbose) {
          console.warn(`Warning: (${dir}) already removed`);
        }
        const parentDir = import_pathe.default.dirname(dir);
        if (parentDir !== baseDir) {
          await this.cleanEmptyDir(parentDir, baseDir);
        }
        return;
      }
      throw new FileMoverError(
        `Failed to clean directory`,
        "CLEAN_ERROR",
        dir,
        void 0,
        err
      );
    }
  }
  /**
   * 清理文件或目录
   */
  async cleanFile(source) {
    const paths = Array.isArray(source) ? source : [source];
    await Promise.all(
      paths.map(async (p) => {
        try {
          this._emit("clean:start", p);
          await import_fs_extra.default.remove(p);
          if (this.options.clean) {
            await this.cleanEmptyDir(import_pathe.default.dirname(p), import_pathe.default.resolve(this.options.cwd));
          }
          this._emit("clean:done", p);
        } catch (err) {
          throw new FileMoverError(
            `Failed to clean file`,
            "CLEAN_ERROR",
            p,
            void 0,
            err
          );
        }
      })
    );
  }
  /**
   * 根据映射表移动文件或目录
   */
  async move(pathMap) {
    const { cwd, base, dest, concurrency, verbose } = this.options;
    const basePath = /\w+/.test(base) ? import_pathe.default.resolve(cwd, base) : null;
    const entries = Object.entries(pathMap);
    const chunks = Array.from(
      { length: Math.ceil(entries.length / concurrency) },
      (_, i) => entries.slice(i * concurrency, (i + 1) * concurrency)
    );
    for (const chunk of chunks) {
      await Promise.all(
        chunk.map(async ([source, target]) => {
          const sourcePath = import_pathe.default.resolve(cwd, source);
          const targetPath = import_pathe.default.resolve(cwd, dest, target);
          let subPath = "";
          try {
            const sourceStats = await import_fs_extra.default.stat(sourcePath);
            if (basePath && sourcePath.startsWith(basePath)) {
              subPath = import_pathe.default.relative(basePath, sourcePath);
            }
            await this.copyFile(sourcePath, targetPath, sourceStats.isDirectory(), subPath);
            await this.cleanFile(sourcePath);
          } catch (err) {
            if (err.code === "ENOENT") {
              const files = await (0, import_fast_glob.default)(source, {
                cwd,
                absolute: true
                // onlyFiles: false,
              });
              if (files.length === 0) {
                if (verbose) {
                  console.warn(`Warning: No files found matching pattern ${source}`);
                }
                return;
              }
              await Promise.all(
                files.map(async (file) => {
                  if (basePath && file.startsWith(basePath)) {
                    subPath = import_pathe.default.relative(basePath, file);
                  }
                  await this.copyFile(file, targetPath, true, subPath);
                })
              );
              await this.cleanFile(files);
            } else {
              throw new FileMoverError(
                `Failed to process file`,
                "PROCESS_ERROR",
                sourcePath,
                targetPath,
                err
              );
            }
          }
        })
      );
    }
  }
};
function createFileMover(options) {
  return new FileMover(options);
}
async function moveFile(pathMap, options = {}) {
  const mover = new FileMover(options);
  await mover.move(pathMap);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FileMoverError,
  createFileMover,
  moveFile
});
