// mv-file/dist/index.mjs
// src/index.ts
import fs from "fs-extra";
import path from "pathe";
import fg from "fast-glob";
import { EventEmitter } from "events";
var FileMoverError = class extends Error {
  constructor(message, code, source, target, originalError) {
    super(message);
    this.code = code;
    this.source = source;
    this.target = target;
    this.originalError = originalError;
    this.name = "FileMoverError";
  }
};
var FileMover = class extends EventEmitter {
  constructor(options = {}) {
    super();
    this.options = {
      cwd: process.cwd(),
      base: "",
      dest: "",
      force: false,
      clean: false,
      verbose: false,
      concurrency: 4,
      ...options
    };
  }
  _emit(event, ...args) {
    return super.emit(event, ...args);
  }
  /**
   * 复制文件或目录
   * @param source 源文件路径
   * @param target 目标路径
   * @param isDir 源路径是否为目录
   * @param subPath 附加路径(相对base的路径)
   */
  async copyFile(source, target, isDir, subPath) {
    try {
      let finalTarget = target;
      const { force } = this.options;
      if (isDir) {
        if (subPath) finalTarget = path.join(target, subPath);
      } else {
        const sourceExt = path.extname(source);
        const targetExt = path.extname(target);
        if (!targetExt && sourceExt !== targetExt) {
          finalTarget = subPath ? path.join(target, subPath) : path.join(target, path.basename(source));
        }
      }
      this._emit("copy:start", source, finalTarget);
      await fs.copy(source, finalTarget, {
        overwrite: force,
        errorOnExist: !force,
        preserveTimestamps: true
      });
      this._emit("copy:done", source, finalTarget);
    } catch (err) {
      throw new FileMoverError(`Failed to copy file`, "COPY_ERROR", source, target, err);
    }
  }
  /**
   * 清理空目录
   * @param {string} dir - 清理目录
   * @param {string} baseDir - 基准目录
   */
  async cleanEmptyDir(dir, baseDir = "") {
    try {
      if (!dir.startsWith(baseDir)) return;
      const items = await fs.readdir(dir);
      if (items.length > 0) return;
      await fs.rmdir(dir);
      const parentDir = path.dirname(dir);
      if (parentDir !== baseDir) {
        await this.cleanEmptyDir(parentDir, baseDir);
      }
    } catch (err) {
      if (err.code === "ENOENT") {
        if (this.options.verbose) {
          console.warn(`Warning: (${dir}) already removed`);
        }
        const parentDir = path.dirname(dir);
        if (parentDir !== baseDir) {
          await this.cleanEmptyDir(parentDir, baseDir);
        }
        return;
      }
      throw new FileMoverError(
        `Failed to clean directory`,
        "CLEAN_ERROR",
        dir,
        void 0,
        err
      );
    }
  }
  /**
   * 清理文件或目录
   */
  async cleanFile(source) {
    const paths = Array.isArray(source) ? source : [source];
    await Promise.all(
      paths.map(async (p) => {
        try {
          this._emit("clean:start", p);
          await fs.remove(p);
          if (this.options.clean) {
            await this.cleanEmptyDir(path.dirname(p), path.resolve(this.options.cwd));
          }
          this._emit("clean:done", p);
        } catch (err) {
          throw new FileMoverError(
            `Failed to clean file`,
            "CLEAN_ERROR",
            p,
            void 0,
            err
          );
        }
      })
    );
  }
  /**
   * 根据映射表移动文件或目录
   */
  async move(pathMap) {
    const { cwd, base, dest, concurrency, verbose } = this.options;
    const basePath = /\w+/.test(base) ? path.resolve(cwd, base) : null;
    const entries = Object.entries(pathMap);
    const chunks = Array.from(
      { length: Math.ceil(entries.length / concurrency) },
      (_, i) => entries.slice(i * concurrency, (i + 1) * concurrency)
    );
    for (const chunk of chunks) {
      await Promise.all(
        chunk.map(async ([source, target]) => {
          const sourcePath = path.resolve(cwd, source);
          const targetPath = path.resolve(cwd, dest, target);
          let subPath = "";
          try {
            const sourceStats = await fs.stat(sourcePath);
            if (basePath && sourcePath.startsWith(basePath)) {
              subPath = path.relative(basePath, sourcePath);
            }
            await this.copyFile(sourcePath, targetPath, sourceStats.isDirectory(), subPath);
            await this.cleanFile(sourcePath);
          } catch (err) {
            if (err.code === "ENOENT") {
              const files = await fg(source, {
                cwd,
                absolute: true
                // onlyFiles: false,
              });
              if (files.length === 0) {
                if (verbose) {
                  console.warn(`Warning: No files found matching pattern ${source}`);
                }
                return;
              }
              await Promise.all(
                files.map(async (file) => {
                  if (basePath && file.startsWith(basePath)) {
                    subPath = path.relative(basePath, file);
                  }
                  await this.copyFile(file, targetPath, true, subPath);
                })
              );
              await this.cleanFile(files);
            } else {
              throw new FileMoverError(
                `Failed to process file`,
                "PROCESS_ERROR",
                sourcePath,
                targetPath,
                err
              );
            }
          }
        })
      );
    }
  }
};
function createFileMover(options) {
  return new FileMover(options);
}
async function moveFile(pathMap, options = {}) {
  const mover = new FileMover(options);
  await mover.move(pathMap);
}
export {
  FileMoverError,
  createFileMover,
  moveFile
};
