// vite-plugin-vanilla/dist/index.js
// src/index.ts
import createMinifyPlugin from "vite-plugin-minify-html";

// src/vanillaPlugin.ts
import fs from "fs/promises";
import path from "pathe";
import glob from "fast-glob";

// src/utils.ts
import * as vite from "vite";
import { error as errorLog, colors } from "diy-log";
var PLUGIN_NAME = "vite-plugin-vanilla";
var getPageRE = (suffix, end = true) => {
  const suffixStr = Array.isArray(suffix) ? suffix.join("|") : suffix;
  return new RegExp(`\\.(${suffixStr})${end ? "$" : ""}`, "i");
};
function errlog(...args) {
  errorLog(`[${colors.gray(PLUGIN_NAME)}] `, ...args);
}
function getViteVersion() {
  return vite?.version ? Number(vite.version.split(".")[0]) : 2;
}
function cleanUrl(url) {
  if (!url) return "/";
  const queryRE = /\?.*$/s;
  const hashRE = /#.*$/s;
  return url.replace(hashRE, "").replace(queryRE, "");
}
function cleanPageUrl(path2, reg) {
  const result = path2.replace(/(^\/)|(\/$)/g, "");
  return reg ? result.replace(reg, "") : result;
}

// src/vanillaPlugin.ts
import { moveFile } from "mv-file";
var defaults = {
  include: "src/**/*.html",
  exclude: [],
  suffix: "html",
  base: "src",
  minify: true,
  // transform: ()=>{},
  inject: { data: {}, tags: [] },
  replaceDefine: true
};
function createVanillaPlugin(options = {}) {
  const htmlPages = {};
  let viteConfig;
  const opts = Object.assign({}, defaults, options);
  const pageRE = getPageRE(opts.suffix);
  const pagePathRE = getPageRE(opts.suffix, false);
  const basePathResolve = path.resolve(opts.base);
  const transformIndexHtmlHandler = async (html, ctx) => {
    try {
      if (opts.replaceDefine && viteConfig.define) {
        for (const key in viteConfig.define) {
          const _value = viteConfig.define[key];
          const keyReg = new RegExp(key.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"), "g");
          html = html.replace(keyReg, typeof _value === "string" ? JSON.parse(_value) : _value);
        }
      }
      if (typeof opts.transform === "function") {
        html = await opts.transform(html, ctx);
      }
    } catch (err) {
      errlog(err.message);
    }
    return {
      html,
      tags: opts.inject?.tags || []
    };
  };
  return {
    name: PLUGIN_NAME,
    configResolved(resolvedConfig) {
      viteConfig = resolvedConfig;
    },
    config(config) {
      const input = {};
      const files = glob.sync(opts.include, { ignore: opts.exclude });
      files.forEach((file) => {
        if (!pageRE.test(file)) return;
        const absolutePath = path.resolve(file);
        const relativePath = path.relative(
          path.join(config?.root ?? process.cwd(), opts.base),
          absolutePath
        );
        const fileUrl = cleanPageUrl(relativePath, pageRE);
        input[fileUrl] = absolutePath;
        htmlPages[fileUrl] = {
          file,
          //文件路径
          path: fileUrl,
          //访问路径
          filePath: absolutePath,
          // 绝对路径
          output: relativePath
          // 输出路径
        };
      });
      return {
        appType: "mpa",
        build: {
          rollupOptions: {
            input
          }
        }
      };
    },
    configureServer(server) {
      server.middlewares.use(async (req, res, next) => {
        let url = cleanUrl(req.url ?? req.originalUrl ?? "");
        const baseDir = viteConfig.base;
        if (baseDir && url.startsWith(baseDir)) {
          url = url.substring(baseDir.length);
        }
        url = cleanPageUrl(url, pageRE);
        if (url === "") {
          url = "index";
        }
        const pageData = pagePathRE.test(req.url || "") ? htmlPages[url] : htmlPages[url] || htmlPages[`${url}/index`];
        if (pageData) {
          try {
            const html = await fs.readFile(pageData.filePath, "utf-8");
            const transformedHtml = await server.transformIndexHtml(
              req.url || `/${url}`,
              html,
              req.originalUrl
            );
            res.setHeader("Content-Type", "text/html");
            return res.end(transformedHtml);
          } catch (err) {
            errlog(err.message);
            return next(err);
          }
        }
        next();
      });
      const watcher = server.watcher;
      Object.values(htmlPages).forEach(({ filePath }) => {
        watcher.add(filePath);
      });
      function handleChangeFile(_file) {
        server.ws.send({
          type: "full-reload",
          path: "*"
        });
      }
      function handleAddFile(file) {
        const _path = path.resolve(file);
        if (_path.includes(basePathResolve) && pageRE.test(_path)) {
          server.restart();
        }
      }
      watcher.on("change", handleChangeFile);
      watcher.on("unlink", handleChangeFile);
      watcher.on("add", handleAddFile);
    },
    transformIndexHtml: getViteVersion() < 5 ? {
      enforce: "pre",
      transform: transformIndexHtmlHandler
    } : {
      order: "pre",
      handler: transformIndexHtmlHandler
    },
    closeBundle() {
      const dest = viteConfig.build.outDir || "dist";
      const files = Object.values(htmlPages).reduce((acc, page) => {
        const _src = path.join(dest, page.file);
        acc[_src] = page.output;
        return acc;
      }, {});
      moveFile(files, { base: dest, dest, clean: true, force: true });
    }
  };
}

// src/index.ts
function createPlugin(options = {}) {
  if (typeof options === "string" || Array.isArray(options)) {
    options = { include: options };
  }
  const opts = Object.assign({ minify: true }, options);
  const plugins = [createVanillaPlugin(opts)];
  if (opts.minify) {
    plugins.push(createMinifyPlugin(opts.minify));
  }
  return plugins;
}
export {
  createPlugin as default
};
